File: server/module/module-erp/erp-server/src/convert/erp_purchase_return.rs
use erp_model::response::erp_purchase_return_attachment::ErpPurchaseReturnAttachmentBaseResponse;
use erp_model::response::erp_purchase_return_detail::ErpPurchaseReturnDetailBaseResponse;
use sea_orm::{Set, NotSet};
use crate::model::erp_purchase_return::{self, Model as ErpPurchaseReturn, ActiveModel as ErpPurchaseReturnActiveModel};
use crate::model::erp_purchase_order::{self, Model as ErpPurchaseOrder};
use crate::model::erp_supplier::{self, Model as ErpSupplier};
use crate::model::erp_settlement_account::{self, Model as ErpSettlementAccount};
use erp_model::request::erp_purchase_return::{CreateErpPurchaseReturnRequest, UpdateErpPurchaseReturnRequest};
use erp_model::response::erp_purchase_return::{ErpPurchaseReturnBaseResponse, ErpPurchaseReturnInfoResponse, ErpPurchaseReturnPageResponse, ErpPurchaseReturnResponse};

pub fn create_request_to_model(request: &CreateErpPurchaseReturnRequest) -> ErpPurchaseReturnActiveModel {
    ErpPurchaseReturnActiveModel {
        purchase_order_id: Set(request.purchase_order_id.clone()),
        return_date: Set(request.return_date.clone()),
        total_amount: Set(request.total_amount.clone()),
        discount_rate: request.discount_rate.as_ref().map_or(NotSet, |discount_rate| Set(Some(discount_rate.clone()))),
        settlement_account_id: request.settlement_account_id.as_ref().map_or(NotSet, |settlement_account_id| Set(Some(settlement_account_id.clone()))),
        deposit: request.deposit.as_ref().map_or(NotSet, |deposit| Set(Some(deposit.clone()))),
        remarks: request.remarks.as_ref().map_or(NotSet, |remarks| Set(Some(remarks.clone()))),
        ..Default::default()
    }
}

pub fn update_request_to_model(request: &UpdateErpPurchaseReturnRequest, existing: ErpPurchaseReturn) -> ErpPurchaseReturnActiveModel {
    let mut active_model: ErpPurchaseReturnActiveModel = existing.into();
    if let Some(return_date) = &request.return_date { 
        active_model.return_date = Set(return_date.clone());
    }
    if let Some(total_amount) = &request.total_amount { 
        active_model.total_amount = Set(total_amount.clone());
    }
    if let Some(order_status) = &request.order_status { 
        active_model.order_status = Set(order_status.clone());
    }
    if let Some(discount_rate) = &request.discount_rate { 
        active_model.discount_rate = Set(Some(discount_rate.clone()));
    }
    if let Some(settlement_account_id) = &request.settlement_account_id { 
        active_model.settlement_account_id = Set(Some(settlement_account_id.clone()));
    }
    if let Some(deposit) = &request.deposit { 
        active_model.deposit = Set(Some(deposit.clone()));
    }
    if let Some(remarks) = &request.remarks { 
        active_model.remarks = Set(Some(remarks.clone()));
    }
    active_model
}

pub fn model_to_response(model: ErpPurchaseReturn) -> ErpPurchaseReturnResponse {
    ErpPurchaseReturnResponse { 
        id: model.id,
        order_number: model.order_number,
        purchase_order_id: model.purchase_order_id,
        supplier_id: model.supplier_id,
        return_date: model.return_date,
        total_amount: model.total_amount,
        order_status: model.order_status,
        discount_rate: model.discount_rate,
        settlement_account_id: model.settlement_account_id,
        deposit: model.deposit,
        remarks: model.remarks,
        department_code: model.department_code,
        department_id: model.department_id,
        creator: model.creator,
        create_time: model.create_time,
        updater: model.updater,
        update_time: model.update_time,
    }
}

pub fn model_to_page_response(model: ErpPurchaseReturn, model_purchase: Option<ErpPurchaseOrder>, model_supplier: Option<ErpSupplier>, model_settlement_account: Option<ErpSettlementAccount>) -> ErpPurchaseReturnPageResponse {
    let purchase_order_number = model_purchase.map(|purchase| purchase.order_number.clone());
    let supplier_name = model_supplier.map(|supplier| supplier.name.clone());
    let settlement_account_name = model_settlement_account.map(|settlement_account| settlement_account.name.clone());

    ErpPurchaseReturnPageResponse { 
        id: model.id,
        order_number: model.order_number,
        purchase_order_id: model.purchase_order_id,
        supplier_id: model.supplier_id,
        return_date: model.return_date,
        total_amount: model.total_amount,
        order_status: model.order_status,
        discount_rate: model.discount_rate,
        settlement_account_id: model.settlement_account_id,
        deposit: model.deposit,
        remarks: model.remarks,
        department_code: model.department_code,
        department_id: model.department_id,
        creator: model.creator,
        create_time: model.create_time,
        updater: model.updater,
        update_time: model.update_time,

        purchase_order_number,
        supplier_name,
        settlement_account_name
    }
}

pub fn model_to_base_response(model: ErpPurchaseReturn, details: Vec<ErpPurchaseReturnDetailBaseResponse>, attachments: Vec<ErpPurchaseReturnAttachmentBaseResponse>) -> ErpPurchaseReturnBaseResponse {
    ErpPurchaseReturnBaseResponse { 
        id: model.id,
        order_number: model.order_number,
        purchase_order_id: model.purchase_order_id,
        supplier_id: model.supplier_id,
        return_date: model.return_date,
        total_amount: model.total_amount,
        order_status: model.order_status,
        discount_rate: model.discount_rate,
        settlement_account_id: model.settlement_account_id,
        deposit: model.deposit,
        remarks: model.remarks,

        details,
        attachments,
    }
}

pub fn model_to_info_response(model: ErpPurchaseReturn, model_settlement_account: Option<ErpSettlementAccount>, details: Vec<ErpPurchaseReturnDetailBaseResponse>, attachments: Vec<ErpPurchaseReturnAttachmentBaseResponse>) -> ErpPurchaseReturnInfoResponse {
    let settlement_account_name = model_settlement_account.map(|settlement_account| settlement_account.name.clone());

    ErpPurchaseReturnInfoResponse { 
        id: model.id,
        order_number: model.order_number,
        purchase_order_id: model.purchase_order_id,
        supplier_id: model.supplier_id,
        return_date: model.return_date,
        total_amount: model.total_amount,
        order_status: model.order_status,
        discount_rate: model.discount_rate,
        settlement_account_id: model.settlement_account_id,
        deposit: model.deposit,
        remarks: model.remarks,

        details,
        attachments,
        settlement_account_name
    }
}

File: admin/src/api/erp_purchase_return.ts
import { PaginatedRequest, PaginatedResponse } from "@/base/page";
import { api } from "@/utils/request";
import { ErpPurchaseReturnDetailBaseResponse, ErpPurchaseReturnDetailRequest } from "./erp_purchase_return_detail";
import { ErpPurchaseReturnAttachmentBaseResponse, ErpPurchaseReturnAttachmentRequest } from "./erp_purchase_return_attachment";

const apis = {
  create: "/erp/erp_purchase_return/create", // 新增
  update: "/erp/erp_purchase_return/update", // 修改
  delete: "/erp/erp_purchase_return/delete", // 删除
  get: "/erp/erp_purchase_return/get", // 单条查询
  get_base: "/erp/erp_purchase_return/get_base", // 单条查询
  get_info: "/erp/erp_purchase_return/get_info", // 单条查询
  list: "/erp/erp_purchase_return/list", // 列表查询
  list_supplier: "/erp/erp_purchase_return/list_supplier", // 列表查询
  page: "/erp/erp_purchase_return/page", // 分页查询
};

export interface ErpPurchaseReturnRequest {
  id?: number; // 退货ID
  purchase_order_id?: number; // 采购订单ID
  return_date: string; // 退货日期
  total_amount: number; // 总金额
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id: number; // 结算账户ID
  deposit?: number; // 定金
  remarks: string; // 备注

  details: ErpPurchaseReturnDetailRequest[]; // 入库采购产品仓库列表
  attachments: ErpPurchaseReturnAttachmentRequest[]; // 入库附件列表
}

export interface ErpPurchaseReturnResponse {
  id: number; // 退货ID
  order_number: number; // 订单编号
  purchase_order_id: number; // 采购订单ID
  supplier_id: number; // 供应商ID
  return_date: string; // 退货日期
  total_amount: number; // 总金额
  order_status: number; // 订单状态
  discount_rate: number; // 优惠率（百分比，1000表示10.00%）
  settlement_account_id: number; // 结算账户ID
  deposit: number; // 定金
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间

  purchase_order_number: number; // 采购订单编号
  supplier_name: string; // 供应商
  settlement_account_name: string; // 结算账户

  details: ErpPurchaseReturnDetailBaseResponse[]; // 入库采购产品仓库列表
  attachments: ErpPurchaseReturnAttachmentBaseResponse[]; // 入库附件列表
}

export interface ErpPurchaseReturnQueryCondition extends PaginatedRequest {}

export const createErpPurchaseReturn = (
  erp_purchase_return: ErpPurchaseReturnRequest
): Promise<number> => {
  return api.post<number>(apis.create, erp_purchase_return);
};

export const updateErpPurchaseReturn = (
  erp_purchase_return: ErpPurchaseReturnRequest
): Promise<void> => {
  return api.post<void>(apis.update, erp_purchase_return);
};

export const deleteErpPurchaseReturn = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
};

export const getErpPurchaseReturn = (
  id: number
): Promise<ErpPurchaseReturnResponse> => {
  return api.get<ErpPurchaseReturnResponse>(`${apis.get}/${id}`);
};

export const getBaseErpPurchaseReturn = (
  id: number
): Promise<ErpPurchaseReturnResponse> => {
  return api.get<ErpPurchaseReturnResponse>(`${apis.get_base}/${id}`);
};

export const getInfoErpPurchaseReturn = (
  id: number
): Promise<ErpPurchaseReturnResponse> => {
  return api.get<ErpPurchaseReturnResponse>(`${apis.get_info}/${id}`);
};

export const listErpPurchaseReturn = (): Promise<
  Array<ErpPurchaseReturnResponse>
> => {
  return api.get<Array<ErpPurchaseReturnResponse>>(apis.list);
};

export const listSupplierErpPurchaseReturn = (supplier_id: number): Promise<
  Array<ErpPurchaseReturnResponse>
> => {
  return api.get<Array<ErpPurchaseReturnResponse>>(`${apis.list_supplier}/${supplier_id}`);
};

export const pageErpPurchaseReturn = (
  condition: ErpPurchaseReturnQueryCondition
): Promise<PaginatedResponse<ErpPurchaseReturnResponse>> => {
  return api.get<PaginatedResponse<ErpPurchaseReturnResponse>>(
    apis.page,
    condition
  );
};
